### ISSUE TITLE
A page showing orgs logs for changes their objects

### ISSUE THREAD

**Issue by leovegoda** [peeringdb/peeringdb#979]
**Is your feature request related to a problem? Please describe.**
Based on a discussion with Guy Tal (Lumen). Large organizations often have many PeeringDB users and they can be located across multiple time zones. When a record changes, finding out who made the change and why can take time. It would be helpful if PeeringDB made change logs available to each org’s admin users.

**Who is affected by the problem?**
Organizations with multiple users of PeeringDB in different reporting chains.

**What is the impact?**
It can be hard to discover who made a change and why in large organizations when many users have access to PeeringDB.

**Are there security concerns?**
No

**Are there privacy concerns?**
No

**Describe the solution you'd like**
Create a page that shows users affiliated with an organization, all changes made for that organization’s records. 

- Timestamp
- username 
- Nature of change

e.g. 

YYYY-MM-DD HH:MM:SS (UTC) username added facility “Name label” “id”
YYYY-MM-DD HH:MM:SS (UTC)  username changed notes from “TEXT1” to “TEXT2”
YYYY-MM-DD HH:MM:SS (UTC) username removed facility “Name label” “id”

**Describe alternatives you've considered**
Allow users to select types of changes to be notified to a specified e-mail address.

**Could this feature request need support from the Admin Committee?**
Possibly. @peeringdb/ac should review.

**What is the proposed priority?**
This calendar year

**Provide a rationale for any/all of the above**
There is no reason that senior users at organizations should not be able to review what has changed for their organization, so they can understand the rationale and either make further changes or communicate internally.

**Additional context**
None

**Comment by peterhelmenstine** [comment]
+1 great idea

**Comment by martinhannigan** [comment]
+1

**Comment by DarwinCosta** [comment]
+1

**Comment by grizz** [comment]
+1

Everything is in place to do this, let's get it done @peeringdb/pc

**Comment by arnoldnipper** [comment]
+1

**Comment by leovegoda** [comment]
There's been strong support for this from the AC. Does anyone else on the @peeringdb/pc want to support it?

**Comment by arnoldnipper** [comment]
> There's been strong support for this from the AC. Does anyone else on the @peeringdb/pc want to support it?

jfyi: this issue is not flagged AC. And AC already has full access to the history of objects. Nevertheless, one more +1 from @peeringdb/pc would be nice :)

**Comment by job** [comment]
+1

**Comment by DarwinCosta** [comment]
+1

**Comment by vegu** [comment]
This should be straight forward to do with the current historical snapshot backend we have - only potential issue that jumps out at me is whether or not this changelog needs to persist when objects are moved to a different organization. 

Do we need to show a history of the object up until to the point where it was moved or are we ok with the object's history disappearing completely once it is moved to a different org.

**Comment by leovegoda** [comment]
> Do we need to show a history of the object up until to the point where it was moved or are we ok with the object's history disappearing completely once it is moved to a different org.

I do not think so. The goal is to inform an organization about changes made to an object they manage to gain an understanding of recent changes by contacting the person responsible. The value of the information will diminish as time passes because people will change and memories will fade. Once the people are in different organizations the value of the information is likely to be tenuous at best.

**Comment by arnoldnipper** [comment]
> Do we need to show a history of the object up until to the point where it was moved or are we ok with the object's history disappearing completely once it is moved to a different org.

The history of the merged org should still be accessible to the new Admins, but kept separate in case we have to do an unmerge.

**Comment by leovegoda** [comment]
I propose we implement this with a 3 month history and then iterate based on user and committee feedback to see if we need to extend the period or add logic for rules like logs that cross organizational changes

**Comment by arnoldnipper** [comment]
> I propose we implement this with a 3 month history

What is the logic to limit this to three months?

**Comment by martinhannigan** [comment]
Should we wrap a little more detail in?

who can see the change logs?
how long do we archive?
are they exportable via API?
i assume costs are analyzed?
noc or SEC standard format requirement?
configurable settings eg time?

will review privacy  policy, but we review for compliance?

what else?

Thanks

**Comment by leovegoda** [comment]
> > I propose we implement this with a 3 month history
> 
> What is the logic to limit this to three months?

Human memory. A log that refers to an event from a year ago or person who has left the organization is less useful. If we wanted to offer additional value we could allow users to enter an internal reference, like a ticket number, alongside any change they make in PeeringDB, and include that in the log.

But I'd go with a simple implementation at first and iterate based on user feedback rather than try to mke the original design perfect.

**Comment by martinhannigan** [comment]
On Thu, Sep 23, 2021 at 9:04 AM Leo Vegoda ***@***.***> wrote:

> I propose we implement this with a 3 month history
>
> What is the logic to limit this to three months?
>
>
It would be helpful to quote. If this is a github interface limit,
apologies.

Human memory. A log that refers to an event from a year ago or person who
> has left the organization is less useful. If we wanted to offer additional
> value we could allow users to enter an internal reference, like a ticket
> number, alongside any change they make in PeeringDB, and include that in
> the log.
>

You don't know until you know. I would hope for some ability to default to
something low and allow it to be longer "somehow".

But I'd go with a simple implementation at first and iterate based on user
> feedback rather than try to mke the original design perfect.
>

That makes sense.


Warm regards,

-M<

**Comment by arnoldnipper** [comment]
> But I'd go with a simple implementation at first and iterate based on user feedback rather than try to mke the original design perfect.

For me, a simple implementation show unlimited history. If you limit, this means extra code.

**Comment by arnoldnipper** [comment]
**Summary**
For most of the objects, an internal changelog is kept. However, this changelog is not visible to the owners.
- make all changelogs accessible (i.e. no time constraint)
- users with r/w permissions are able to read the logs (i.e. if  a user is given permission to admin a `fac`, make the log for this `fac` accessible to this user)
- if two objects are merged, keep the logs distinguishable so that in case of an unmerge the logs also can be unmerged

@peeringdb/pc and @peeringdb/ac, please make up your mind. Same for @peeringdb/20c

**Comment by mcmanuss8** [comment]
> **Summary** For most of the objects, an internal changelog is kept. However, this changelog is not visible to the owners.
> 
> * make all changelogs accessible (i.e. no time constraint)
> * users with r/w permissions are able to read the logs (i.e. if  a user is given permission to admin a `fac`, make the log for this `fac` accessible to this user)
> * if two objects are merged, keep the logs distinguishable so that in case of an unmerge the logs also can be unmerged
> 
> @peeringdb/pc and @peeringdb/ac, please make up your mind. Same for @peeringdb/20c

+1

**Comment by arnoldnipper** [comment]
@mcmanuss8, is [Add a log/audit trail for all objects to API and Web] (#915) a separate issue?

**Comment by mcmanuss8** [comment]
#915 is more of a global/public view. Let's start with this issue as-is and deal with #915 separately. I'll add comments into #915 clarifying the difference between this issue and that

**Comment by arnoldnipper** [comment]
> **Are there privacy concerns?**
> No

Yes, there are. What do we do when a user removes their account?

**Comment by leovegoda** [comment]
> Putting back to 3a because of the privacy concerns

Can you state what the actual concerns are? If a person works for an organization and makes a change to a configuration in a system they run, they wouldn't expect the organization to remove the log of their change when they left the organization. Organizations have a legitimate interest in knowing which of their staff or contractors make changes and when

**Comment by arnoldnipper** [comment]
> Can you state what the actual concerns are? If a person works for an organization and makes a change to a configuration in a system they run, they wouldn't expect the organization to remove the log of their change when they left the organization. Organizations have a legitimate interest in knowing which of their staff or contractors make changes and when

Sure! However, when a user removes their account, there is GDPR on one side and auditing trails on the other. IMO GDPR wins

**Comment by leovegoda** [comment]
> > Can you state what the actual concerns are? If a person works for an organization and makes a change to a configuration in a system they run, they wouldn't expect the organization to remove the log of their change when they left the organization. Organizations have a legitimate interest in knowing which of their staff or contractors make changes and when
> 
> Sure! However, when a user removes their account, there is GDPR on one side and auditing trails on the other. IMO GDPR wins

Can you be specific about what in GDPR prevents an organization from retaining and viewing these logs?

**Comment by martinhannigan** [comment]
On Mon, Aug 1, 2022 at 11:08 Arnold Nipper ***@***.***> wrote:

> *Are there privacy concerns?*
> No
>
> Yes, there are. What do we do when a user removes their account?
>

I don't see the problem.

https://docs.peeringdb.com/gov/misc/2017-04-02-PeeringDB_Privacy_Policy.pdf

Is there another privacy policy somewhere?

Thanks

**Comment by arnoldnipper** [comment]
> Can you be specific about what in GDPR prevents an organization from retaining and viewing these logs?

Nothing. However, personal related data must be removed. This regularly includes usernames if they have been removed

**Comment by leovegoda** [comment]
@gmazoyer and I discussed https://github.com/peeringdb/peeringdb/issues/1328 and agreed that it might be helpful to add the ability to leave a comment with a change, that becomes visible to organizational admins in this log. The comment could be some plain text or a reference number that is meaningful to the organization. Thoughts from the @peeringdb/pc ?

**Comment by grizz** [comment]
I like the plain text comment field
